functions/config-versioning.bash

```diff
@@ -0,0 +1,310 @@
+    then
+      echo "OK"
+    else 
+      echo "CANCELED"
+      return 0 
+  fi
+ 
+  echo "$logPrefix Stopping openhab"; systemctl stop openhab.service
+  echo -n "$logPrefix Gathering Credentials for Github..."
+
+    if [ -z "$user" ]
+    then
+      if ! user="$(whiptail --title "Versioning Github" --inputbox "\\nGithub Username?" 8 39 --title "$questionText Username" 3>&1 1>&2 2>&3)"; then
+        echo "CANCELED UN"; return 0
+      fi
+    fi
+
+    if [ -z "$email" ]
+    then
+      if ! email="$(whiptail --title "Versioning Github" --inputbox "\\nGithub Email?" 8 39 --title "$questionText Email" 3>&1 1>&2 2>&3)"; then
+        echo "CANCELED UN"; return 0
+      fi
+    fi
+
+    if [ -z "$repositiory" ] 
+    then
+      if ! repositiory="$(whiptail --title "Versioning Github" --inputbox "\\nGithub Repository?" 8 39 --title "$questionText Repository" 3>&1 1>&2 2>&3)"; then
+        echo "CANCELED REPO"; return 0
+      fi
+    fi
+
+    if [ -z "$token" ]
+    then
+      if ! token="$(whiptail --title "Versioning Github" --inputbox "\\nGithub token?" 8 39 --title "$questionText Token" 3>&1 1>&2 2>&3)"; then
+        echo "CANCELED Token"; return 0
+      fi
+    fi
+    echo " OK"
+  fi
+   
+  echo -n "$logPrefix Ensuring network connectivity... "
+  if ! running_in_docker && tryUntil "ping -c1 8.8.8.8 &> /dev/null || curl --silent --head https://github.com/ |& grep -qs 'HTTP/1.1 200 OK'" 5 1
+  then
+    echo "FAILED (Can't reach github)"
+    return 0
+  else
+    echo " OK"
+  fi
+  
+  repourl="https://$user:$token@github.com/$user/$repositiory.git"
+  echo -n "$logPrefix Check if $repositiory Exists and credentials valid..."
+  if git ls-remote "$repourl" | grep "HEAD" > /dev/null 
+  then
+    echo " OK"
+  else
+    echo " FAIL"
+      if [[ -n $INTERACTIVE ]] 
+      then 
+        whiptail --title "Operation Fail" --msgbox "Fail invalid credentials\\nYou entered\\n\\nUser         $user\\nEmail         $email\\nRepositiory  $repositiory\\nToken        $token" 15 80
+      fi
+    return 0
+  fi
+
+  echo -n "$logPrefix Check if $repositiory is set to Private..."
+  if curl -s --head --request GET https://github.com/"$user"/"$repositiory" | grep "HTTP/2 404" > /dev/null
+  then
+    echo " OK"
+  else
+    echo " FAIL"
+    if [[ -n $INTERACTIVE ]]
+    then
+      whiptail --title "Openhabian Versioning" --msgbox "Please Set Repository [$repositiory] to Private on Github and try again" 8 78
+      return 0
+    fi
+  fi
+ 
+  apiurl="https://api.github.com/repos/$user/$repositiory/contents/openhabian_config_versioning"
+  echo -n "$logPrefix Check online Repository [$repositiory] contains previous install..."
+
+  if curl -s -H "Accept: application/vnd.github.v3+json" -i -u "$user":"$token" "$apiurl" | grep "HTTP/1.1 200 OK" > /dev/null
+  then
+    echo -n " YES" # using update_git_repo() caused git to delete everything
+    if ! cond_redirect git -C "/" init; then echo "FAILED (init)"; return 1; fi
```

Comment:
Yep this is an awful situation.  /var/lib/openhab/jsondb is what I want to version the most. To set it up in a different folder and changing the --global core.worktree to root may be better. I can't find or think of another way to do it. If it was just etc then etckeeper would have been the easy solution.

Reply:
Hmm. Use a filesystem link ? If we only ever retrieve/push from within openhabian-config we can make this 2 stages, i.e. a "pull A" would be implemented as "pull B, copy B to A".
Yeah not elegant.

Reply:
That's an excellent thought to use symlink just put it in the /etc/ folder to /var/lib/openhab/jsondb/  I thought the exact same thing.  I tested theory and the file link is versioned but everything through it is ignored. If only it were that easy.

Another Idea is to use etckeeper for etc and have another repo for jsondb that would work. Then the user that wants openHAB 3 to config through UI only need the "openHAB3" Versioning. Users that insist on config files then can install etc keeper.

I will let you guys guide me as to what you ultimately want to achieve as a team.  I think putting git in / is the worst thing about this and it should be put somewhere else maybe /var/lib/ 

It could work the same as etckeeper and preserve permissions.

All is not wasted as knowledge gained is worth my time.

Reply:
> That's an excellent thought to use symlink just put it in the /etc/ folder

have you tried with a hard link, too ? It could link to /, be placed anywhere, used with git -C, and behave like a virtual "/"